### PR TITLE
Add type inference for CastOp

### PR DESCRIPTION
**Pull request by tungld**
This patch is to add a custom type inference for CastOp since the result type of CastOp is specified by one of its attributes, not its inputs.

### CODE REVIEW

**Comment by tjingrant** [comment]
Seems like this patch is related to code here:
https://github.com/onnx/onnx-mlir/blob/20dd6544aae8fefd0c50c2048eb87b9260a0263c/src/Builder/FrontendDialectTransformer.cpp#L60

Any chance we can unify these two?

File: utils/gen_onnx_mlir.py [hunk 1]
```diff
@@ -279,6 +279,38 @@
...
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getIntegerType(/*width=*/64)));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_BOOL*/9:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getI1Type()));
+
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_STRING*/8:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_COMPLEX64*/14:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_COMPLEX128*/15:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_UNDEFINED*/0:
+        default:
+          resultTypes.push_back(builder.getNoneType());
```
Comment (chentong319) [inline utils/gen_onnx_mlir.py:313]: Alternative solution is to give attribute to() to getTypeMap.  Suppose the we should have
`static std::vector<int> getTypeMap() {
      retune {mapping_from_protobuf_to_onnx_mlir(to().getSExtValue())}
}`
Can you manually try that? You need to provide mapping_from_protobuf_to_onnx_mlir, which logically is equivalent to what Tian said about reuse the code.
The solution you provide requires less change in gen_onnx_mlir.py. We may come back to this issue after a unified type inference is provided.

Comment (tungld) [reply utils/gen_onnx_mlir.py:313]: It seems that we cannot call `to()` inside `getTypeMap` since `getTypeMap` is a static method.

Comment (chentong319) [reply utils/gen_onnx_mlir.py:313]: You are right.
I am OK with your current implementation even though it has the issue mentioned by @tjingrant .
To accomplish code reuse as Tian mentioned, I am not clear about where that function should be defined. I tried a little bit and ran into link error.

Comment (tungld) [reply utils/gen_onnx_mlir.py]: I just pushed a commit to make the onnx-to-mlir type translation shareable. Many CMake files are updated to correctly link libraries. If you have any suggestions to reduce CMake updates, please let me know.

**Pull request merged**